File: app/src/main/java/be/digitalia/fosdem/viewmodels/LiveViewModel.java
package be.digitalia.fosdem.viewmodels;

import android.app.Application;
import android.text.format.DateUtils;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Transformations;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;
import be.digitalia.fosdem.db.AppDatabase;
import be.digitalia.fosdem.livedata.LiveDataFactory;
import be.digitalia.fosdem.model.StatusEvent;

import java.util.concurrent.TimeUnit;

public class LiveViewModel extends AndroidViewModel {

	static final long NEXT_EVENTS_INTERVAL = 30L * DateUtils.MINUTE_IN_MILLIS;

	private final AppDatabase appDatabase = AppDatabase.getInstance(getApplication());
	private final LiveData<Long> heartbeat = LiveDataFactory.interval(1L, TimeUnit.MINUTES);
	private final LiveData<PagedList<StatusEvent>> nextEvents = Transformations.switchMap(heartbeat,
			version -> {
				final long now = System.currentTimeMillis();
				return new LivePagedListBuilder<>(appDatabase.getScheduleDao().getEventsWithStartTime(now, now + NEXT_EVENTS_INTERVAL), 20)
						.build();
			});
	private final LiveData<PagedList<StatusEvent>> eventsInProgress = Transformations.switchMap(heartbeat,
			version -> {
				final long now = System.currentTimeMillis();
				return new LivePagedListBuilder<>(appDatabase.getScheduleDao().getEventsInProgress(now), 20)
						.build();
			});

	public LiveViewModel(@NonNull Application application) {
		super(application);
	}

	public LiveData<PagedList<StatusEvent>> getNextEvents() {
		return nextEvents;
	}

	public LiveData<PagedList<StatusEvent>> getEventsInProgress() {
		return eventsInProgress;
	}
}

File: app/src/main/java/be/digitalia/fosdem/viewmodels/LiveViewModel.kt
package be.digitalia.fosdem.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import androidx.paging.PagingSource
import androidx.paging.cachedIn
import be.digitalia.fosdem.db.ScheduleDao
import be.digitalia.fosdem.flow.countSubscriptionsFlow
import be.digitalia.fosdem.flow.stateFlow
import be.digitalia.fosdem.flow.synchronizedTickerFlow
import be.digitalia.fosdem.model.StatusEvent
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.map
import java.time.Duration
import java.time.Instant
import javax.inject.Inject
import kotlin.time.Duration.Companion.minutes

@HiltViewModel
class LiveViewModel @Inject constructor(scheduleDao: ScheduleDao) : ViewModel() {

    // Share a single ticker providing the time to ensure both lists are synchronized
    private val ticker: Flow<Instant> = stateFlow(viewModelScope, null) {
        synchronizedTickerFlow(REFRESH_PERIOD)
            .map { Instant.now() }
    }.filterNotNull()

    @OptIn(ExperimentalCoroutinesApi::class)
    private fun createLiveEventsHotFlow(
        pagingSourceFactory: (now: Instant) -> PagingSource<Int, StatusEvent>
    ): Flow<PagingData<StatusEvent>> {
        return countSubscriptionsFlow {
            ticker
                .flowWhileShared(SharingStarted.WhileSubscribed())
                .distinctUntilChanged()
                .flatMapLatest { now ->
                    Pager(PagingConfig(20)) { pagingSourceFactory(now) }.flow
                }.cachedIn(viewModelScope)
        }
    }

    val nextEvents: Flow<PagingData<StatusEvent>> = createLiveEventsHotFlow { now ->
        scheduleDao.getEventsWithStartTime(now, now + NEXT_EVENTS_INTERVAL)
    }

    val eventsInProgress: Flow<PagingData<StatusEvent>> = createLiveEventsHotFlow { now ->
        scheduleDao.getEventsInProgress(now)
    }

    companion object {
        private val REFRESH_PERIOD = 1.minutes
        private val NEXT_EVENTS_INTERVAL = Duration.ofMinutes(30L)
    }
}